packages/aws-cdk-lib/aws-dynamodb/lib/global-table-base.ts

```diff
@@ -0,0 +1,488 @@
+import { DynamoDBMetrics } from './dynamodb-canned-metrics.generated';
+import * as perms from './perms';
+import { Operation, SystemErrorsForOperationsMetricOptions, OperationsMetricOptions, ITable } from './shared';
+import { IMetric, MathExpression, Metric, MetricOptions, MetricProps } from '../../aws-cloudwatch';
+import { Grant, IGrantable } from '../../aws-iam';
+import { IKey } from '../../aws-kms';
+import { Aws, Resource } from '../../core';
+
+export interface IGlobalTable extends ITable {
+  /**
+   * The ID of the Replica Table in the deployment region of the Global Table.
+   *
+   * @attribute
+   */
+  readonly tableId?: string;
+}
+
+/**
+ * Base class for a Global Table.
+ */
+export abstract class GlobalTableBase extends Resource implements IGlobalTable {
+  /**
+   * The ARN of the Replica Table in the deployment region of the Global Table.
+   *
+   * @attribute
+   */
+  public abstract readonly tableArn: string;
+
+  /**
+   * The name of all Replica Tables in the Global Table.
+   *
+   * @attribute
+   */
+  public abstract readonly tableName: string;
+
+  /**
+   * The stream ARN of the Replica Table in the deployment region of the Global Table.
+   *
+   * @attribute
+   */
+  public abstract readonly tableStreamArn?: string;
+
+  /**
+   * The ID of the Replica Table in the deployment region of the Global Table.
+   *
+   * @attribute
+   */
+  readonly tableId?: string;
+
+  /**
+   * The KMS encryption key associated with the Replica Table in the deployment region
+   * of the Global Table.
+   */
+  public abstract readonly encryptionKey?: IKey;
+
+  protected abstract readonly region: string;
+
+  protected readonly replicaKeys: { [region: string]: IKey } = {};
+
+  protected readonly replicaArns: string[] = [];
+
+  protected readonly streamArns: string[] = [];
+
+  protected abstract get hasIndex(): boolean;
+
+  /**
+   * Adds an IAM policy statement associated with this Global Table to an IAM principal's policy.
+   *
+   * Note: Appropriate grants will also be added to the customer-managed KMS keys associated with this
+   * Global Table and its Replicas.
+   *
+   * @param grantee the principal (no-op if undefined)
+   * @param actions the set of actions to allow (i.e., 'dynamodb:PutItem', 'dynamodb:GetItem', etc.)
+   */
+  public grant(grantee: IGrantable, ...actions: string[]): Grant {
+    return Grant.addToPrincipal({
+      grantee,
+      actions,
+      resourceArns: [
+        this.tableArn,
+        this.hasIndex ? `${this.tableArn}/index/*` : Aws.NO_VALUE,
+        ...this.replicaArns,
```

Comment:
When we do `grantXxx` on a GlobalTable, do we want to grant to all the replicas at the same time? I was thinking we grant only to the primary replica, just like a normal Table

Reply:
I went back and forth with this. This is the current behavior that is used on `Table`. So if you're providing replicas on a `Table` the grants will apply to all replicas and their indexes. I was thinking from the user perspective that might be what is expected since we already do it with a `Table`. My thinking was that you could use `replica` to retrieve the specific replica table in a region and then `grantXxx` there would apply to just the individual replica.

Reply:
After doing some more thinking on this, I think the approach we should proceed with (as you mentioned) is that `grantXxx` should only apply to the primary replica. This issue (https://github.com/aws/aws-cdk/pull/7453) was what led to `grantXxx` for a `Table` applying to all replicas. However, `Table` offered no functionality allowing a user to get a specific replica from their `Table` and it appears `grantXxx` applying to all replicas was the only solution. With `GlobalTable` we now offer a method named `replica` that will give the user the ability to `grant` on any of their replicas. I've made the appropriate changes for this. Also, this aligns the behavior of `metricXxx` with `grantXxx` which will give a better experience, more consistent experience.